### PR TITLE
[Compatibility] Add Module#const_added

### PR DESCRIPTION
**Pull request by itarato**
Source: https://github.com/oracle/truffleruby/issues/3039

Module#const_added has been added. [[Feature #17881](https://bugs.ruby-lang.org/issues/17881)]

### CODE REVIEW

File: spec/ruby/core/module/const_added_spec.rb [hunk 1]
```diff
@@ -121,5 +121,22 @@ class SubClass
...
+        end
+      end
+
+      mod.module_eval(<<-RUBY, __FILE__, __LINE__ + 1)
+        TEST = 123
+        TEST = 456
+      RUBY
+
+      ScratchPad.recorded.should == [123, 456]
+    end
```
Comment (andrykonchin) [inline spec/ruby/core/module/const_added_spec.rb:140]: **suggestion:** It seems the callback is called every time a constant is re-assigned a new value. It makes sense to have a separate test case to illustrate this behaviour.

Comment (itarato) [reply spec/ruby/core/module/const_added_spec.rb:140]: Sure thing

**Review by andrykonchin — APPROVED** [review]

**Comment by eregon** [comment]
I worry that this feature, which causes an extra method call for every constant assignment can significantly slow down loading code.
I'm thinking we could optimize e.g. by having a global `boolean constAddedDefined` in RubyContext, and only call if there ever was a definition of `const_added` (not counting the empty core one).

Also we'd be able to remove the `currentNode != null && coreLibrary != null && coreLibrary.isLoaded()` check and just use that boolean instead which seems nice.
Maybe @andrykonchin can do that to speed up the integration of this PR?

**Comment by itarato** [comment]
> I worry that this feature, which causes an extra method call for every constant assignment, can significantly slow down loading code. I'm thinking we could optimize e.g. by having a global `boolean constAddedDefined` in RubyContext, and only call if there ever was a definition of `const_added` (not counting the empty core one).
> 
> Also we'd be able to remove the `currentNode != null && coreLibrary != null && coreLibrary.isLoaded()` check and just use that boolean instead which seems nice. Maybe @andrykonchin can do that to speed up the integration of this PR?
> 
> Longer-term when we have vtable for methods we could use that to quickly check if `const_added` on a given module is the core one and in that case do nothing.

Sure thing, I let you make that change. Thanks!

**Comment by andrykonchin** [comment]
Merged in https://github.com/oracle/truffleruby/commit/cd498f7451fa0456aabe36889d316be88a30b35f

**Pull request closed without merging**